### ISSUE TITLE
Collector Infusion Enchantment

### ISSUE THREAD

**Issue by DoomRater** [Focamacho/DupeFix-Project#6]
I have a modpack with Thaumcraft 6 and Ex Nihilio Creatio, and I've discovered some wonkiness when using the Collector Infusion Enchantment on the hammers.

The hammer uses a fallback code in case it breaks something that it has no transformation recipe on.  It seems that trigger gets called by Collector even if it breaks something that it transforms, and the collector infusion collects the original block drop even if the block gets successfully destroyed.  I'm pretty sure this is due to how Thaumcraft intercepts item drops- you can see that a zombie that drops zombie brains will not collect the zombie brains if one were to drop from a zombie after killing one, but all the other drops will be collected, including essences that are dropped and rare drops like iron ingots.

TL;DR: hammers with Collector duplicate the broken block while also dropping the transformed item.  Do you think there is anything you could do on your end to help?

**Comment by Focamacho** [comment]
Hey, I tried to fix the dupe and I would like you to test it to confirm that everything is ok.
Here is the new version: 
[DupeFix Project 1.5.zip](https://github.com/Focamacho/DupeFix-Project/files/4273915/DupeFix.Project.1.5.zip)

**Comment by DoomRater** [comment]
Singleplayer seems to work fine.  I tested on a test server instance, and after realizing my scripts were out of sync with my actual server, it seems that it works on my test server as well.  Guess it's time to take it live and see if it works there too.

Additionally, I took it for a spin in a test pack where Ex Nihilio Creatio wasn't present to make sure it didn't throw issues there.  Seems like it works to me!

**Comment by Focamacho** [comment]
Great!
So I'll upload the new version to CurseForge.
Thanks for your report and testing!

**Comment by DoomRater** [comment]
It turns out this is more of a problem with Collector on anything that transforms an item after the fact, meaning axes with Collector infusion and Scorching Heat enchantment from Astral Sorcery will also duplicate the log being smelted.  I presume pickaxes with Collector that have Scorching Heat will also collect a block that isn't transformed and drop a transformed version of the block, although I haven't tried it yet.

In addition, this fix interferes with the gauntlet's harvesting when it's used from the off hand while an item with collector is held.  I have to remember not to break blocks with my breaking focus while holding my hammer or I don't collect any drops.  This occurs only for blocks that get transformed after the fact.

**Comment by Focamacho** [comment]
> It turns out this is more of a problem with Collector on anything that transforms an item after the fact, meaning axes with Collector infusion and Scorching Heat enchantment from Astral Sorcery will also duplicate the log being smelted. I presume pickaxes with Collector that have Scorching Heat will also collect a block that isn't transformed and drop a transformed version of the block, although I haven't tried it yet.
Hmm.. So I'll need to make a check to any tool with Collector enchantment, but I'm not sure how to do it.
I think Thaumcraft is collecting the loot before others mods can edit it. In ExNihilo Fix I used `exnihilocreatio.registries.manager.ExNihiloRegistryManager.HAMMER_REGISTRY.isRegistered(event.getState())` to check if there is any recipe using the breaked block, if true, I remove all the drops from the block(first than thaumcraft) and ExNihilo just add the new drops.
I don't know how to do it with other tools, maybe checking if exists a furnace recipe using the block? It will fix issues with Scorching Heat enchantment or Auto-Smelt enchantments, but I'm not sure if it's the best way to fix it.

> In addition, this fix interferes with the gauntlet's harvesting when it's used from the off hand while an item with collector is held. I have to remember not to break blocks with my breaking focus while holding my hammer or I don't collect any drops. This occurs only for blocks that get transformed after the fact.
My mistake. I'm always checking the main hand in code, but I forgot the offhand can break blocks too.

**Comment by DoomRater** [comment]
Short of making Collector wait to grab all the modified drops and then apply the collection affect to them, that might be the way to go.  A big thank you for continuing to take a look into it!

**Comment by Focamacho** [comment]
Yeah, it would be the best way, but I'm not sure how to "delay" a event from another mod, since I never did it before and Thaumcraft isn't open source.

**Comment by Focamacho** [comment]
Hey, I changed several things in the fix, and now instead of checking the block for conflicting things, I'm overriding the Collector Enchantment with my own version.
Every time you use a tool containing the Thaumcraft Collector's Enchantment, the enchantment will be removed and my version will be added to the item.
It is the same as the normal enchantment, but it is executed at the right times.
I believe that this can also fix problems with items not being collected by the enchantment.

If possible, can you test the new version for me?
[DupeFix Project 1.6.zip](https://github.com/Focamacho/DupeFix-Project/files/4304869/DupeFix.Project.1.6.zip)

**Comment by DoomRater** [comment]
I've started initial testing.  So far, I am noticing that zombie brains are being collected, which is quite nice.  But that is a single player test in my testing modpack.  Next, to test in my main modpack, both singleplayer and server...  I must say I am super excited to see how Collector works.

Unfortunately, when I got to my world test, which has previously infusion enchanted tools, I was still able to collect dupliated items with a hammer.  I attempted to make a new hammer with collector which appears to work, even when I give it similar enchantments.  I think I'm the only player on the server with these sorts of godlike tools, so I can just remake them, but something about the Collector replacement isn't quite working.

**Comment by Focamacho** [comment]
This is strange, because even in the newly created tools I am making the same replacement.

**Comment by DoomRater** [comment]
Sorry for taking so long to get back to here.  I think I see the problem:
https://github.com/Focamacho/DupeFix-Project/blob/a40501a6c5bc64fe614cd892bafc2b63cdac2647/src/main/java/com/focamacho/dupefixproject/fixes/ThaumcraftFixes.java#L67
You loop through all the tags but only ever check the first entry.  The items that never got replaced all have Collector as their last infusion enchantment.

**Comment by Focamacho** [comment]
Ohhhh, my mistake, sorry. I'll fix it now. Thanks for pointing that.

**Comment by Focamacho** [comment]
It'll work now.
[DupeFix Project 1.6.zip](https://github.com/Focamacho/DupeFix-Project/files/4331642/DupeFix.Project.1.6.zip)

**Comment by DoomRater** [comment]
Just confirmed the collector replacement correctly replaced on all my tools the moment they broke something.
